receiver/awsxrayreceiver/internal/udppoller/poller.go

```diff
@@ -0,0 +1,221 @@
+	logger               *zap.Logger
+	wg                   sync.WaitGroup
+	receiverLongLivedCtx context.Context
+	maxPollerCount       int
+	// closing this channel will shutdown all goroutines
+	// within this poller
+	shutDown chan struct{}
+
+	// all segments read by the poller will be sent to this channel
+	segChan chan RawSegment
+}
+
+// New creates a new UDP poller
+func New(cfg *Config, logger *zap.Logger) (Poller, error) {
+	if cfg.Transport != Transport {
+		return nil, fmt.Errorf(
+			"X-Ray receiver only supports ingesting spans through UDP, provided: %s",
+			cfg.Transport,
+		)
+	}
+
+	addr, err := net.ResolveUDPAddr(Transport, cfg.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	sock, err := net.ListenUDP(Transport, addr)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("Listening on endpoint for X-Ray segments",
+		zap.String(Transport, addr.String()))
+
+	return &poller{
+		receiverInstanceName: cfg.ReceiverInstanceName,
+		udpSock:              sock,
+		logger:               logger,
+		maxPollerCount:       cfg.NumOfPollerToStart,
+		shutDown:             make(chan struct{}),
+		segChan:              make(chan RawSegment, segChanSize),
+	}, nil
+}
+
+func (p *poller) Start(receiverLongTermCtx context.Context) {
+	p.receiverLongLivedCtx = receiverLongTermCtx
+	for i := 0; i < p.maxPollerCount; i++ {
+		p.wg.Add(1)
+		go p.poll()
+	}
+}
+
+func (p *poller) Close() error {
+	err := p.udpSock.Close()
+	close(p.shutDown)
+	p.wg.Wait()
+
+	// inform the consumers of segChan that the poller is stopped
+	close(p.segChan)
+	return err
+}
+
+func (p *poller) SegmentsChan() <-chan RawSegment {
+	return p.segChan
+}
+
+func (p *poller) read(buf *[]byte) (int, error) {
+	bufVal := *buf
+	rlen, err := p.udpSock.Read(bufVal)
+	if err == nil {
+		return rlen, nil
+	}
+	switch err := err.(type) {
+	case net.Error:
+		if !err.Temporary() {
+			return -1, fmt.Errorf("read from UDP socket: %w", &recvErr.ErrIrrecoverable{Err: err})
+		}
+	default:
+		return 0, fmt.Errorf("read from UDP socket: %w", &recvErr.ErrRecoverable{Err: err})
+	}
+	return 0, fmt.Errorf("read from UDP socket: %w", &recvErr.ErrRecoverable{Err: err})
+}
+
+func (p *poller) poll() {
+	defer p.wg.Done()
+	buffer := make([]byte, pollerBufferSizeKB)
+
+	var (
+		errRecv   *recvErr.ErrRecoverable
+		errIrrecv *recvErr.ErrIrrecoverable
+	)
+
+	for {
+		select {
+		case <-p.shutDown:
+			return
+		default:
+			// TODO:
+			// call https://pkg.go.dev/go.opentelemetry.io/collector@v0.4.1-0.20200622191610-a8db6271f90a/obsreport?tab=doc#StartTraceDataReceiveOp
+			// once here and
+			// https://pkg.go.dev/go.opentelemetry.io/collector@v0.4.1-0.20200622191610-a8db6271f90a/obsreport?tab=doc#EndTraceDataReceiveOp
+			// at corresponding places in the for loop below.
+			bufPointer := &buffer
+			rlen, err := p.read(bufPointer)
+			if errors.As(err, &errIrrecv) {
+				p.logger.Error("irrecoverable socket read error. Exiting poller", zap.Error(err))
+				return
```

Comment:
Should a new poller be launched if it returns here?

Reply:
Hey @kbrockhoff thanks for the feedback.

I don't think so in this case because if it's an irrecoverable error (i.e. non-transient `net.Error` in `read()` of udppoller), that implies the UDP socket is probably closed or broken to the degree that launching more go routines will not make more progress.

Reply:
The X-Ray daemon takes the approach that if is there is a socket, connection or configuration issue then it kills the entire instance. This approach will not work with the collector because there may be other receivers/exporters that are still working. The collector should not stop processing metrics or logs just because traces are failing.

Tearing down and cleaning up a non-working socket and building a new one is probably the optimum approach for the final state. I am OK with just adding a TODO for now.

To make this work smoothly, we likely need to add a monitor which watches all the receivers/exports. If multiple are failing then the monitor will flip the health status to failing.

Reply:
Love the idea of a health monitor. We can even come up with an interface and force all plugins to implement:
1. health ping/status report
2. best-effort restart when the plugin is unhealthy